x-pack/plugins/security/server/lib/authentication/providers/oidc.js

```diff
@@ -0,0 +1,476 @@
+   * @returns {Promise.<AuthenticationResult>}
+   * @private
+   */
+  async _authenticateViaResponseUrl(request, sessionState) {
+    this._options.log(['debug', 'security', 'oidc'], 'Trying to authenticate via OpenId Connect response query.');
+    // First check to see if this is a Third Party initiated authentication (which can happen via POST or GET)
+    let iss = null;
+    if (request.query) {
+      iss = request.query.iss;
+    } else if (request.payload) {
+      iss = request.payload.iss;
+    }
+    if (iss) {
+      this._options.log(['debug', 'security', 'oidc'], 'Authentication has been initiated by a Third Party.');
+      // We might already have a state and nonce generated by Elasticsearch (from an unfinished authentication in
+      // another tab)
+      const params = this._addStateNonceFromSession({ iss: iss }, sessionState);
+      return this._initiateOpenIdConnectAuthentication(request, params);
+    }
+
+    if (!request.query || !request.query.code) {
+      this._options.log(['debug', 'security', 'oidc'], 'OpenID Connect Authentication response is not found.');
+      return AuthenticationResult.notHandled();
+    }
+
+    // If it is an authentication response and the users' session state doesn't contain all the necessary information,
+    // then something unexpected happened and we should fail because Elasticsearch won't be able to validate the
+    // response.
+    const { nonce: stateNonce, state: stateOidcState, nextURL: stateRedirectURL } = sessionState || {};
+    this._options.log(['debug', 'security', 'oidc-SESSIONSTATE'], sessionState);
+    if (sessionState && (!stateNonce || !stateOidcState || !stateRedirectURL)) {
+      const message = 'Response session state does not have corresponding state or nonce parameters or redirect URL.';
+      this._options.log(['debug', 'security', 'oidc'], message);
+
+      return AuthenticationResult.failed(Boom.badRequest(message));
+    }
+
+    // We have all the necessary parameters, so attempt to complete the OpenID Connect Authentication
+    try {
+      // This operation should be performed on behalf of the user with a privilege that normal
+      // user usually doesn't have `cluster:admin/xpack/security/oidc/authenticate`.
+      const {
+        access_token: accessToken,
+        refresh_token: refreshToken
+      } = await this._options.client.callWithInternalUser(
+        'shield.oidcAuthenticate',
+        // pass the path only as we can't be sure of the full URL and Elasticsearch doesn't need it anyway
+        { body: { state: stateOidcState, nonce: stateNonce, redirect_uri: request.url.path } }
+      );
+
+      this._options.log(['debug', 'security', 'oidc'], 'Request has been authenticated via OpenID Connect.');
+
+      return AuthenticationResult.redirectTo(
+        stateRedirectURL || `${this._options.basePath}/`,
+        { accessToken, refreshToken }
+      );
+    } catch (err) {
+      console.log(
```

Comment:
nit: can we change `console.log` to `this._options.log`?

Reply:
uncleaned debugging:(